src/Functions/FunctionsConversion.h

```diff
@@ -4159,6 +4191,86 @@ arguments, result_type, input_rows_count); \
         };
     }
 
+    template <typename ColumnNumberType, typename EnumType>
+    WrapperType createNumberToEnumWrapper() const
+    {
+        using FieldType = EnumType::FieldType;
+
+        const char * function_name = cast_name;
+        return [function_name] (
+            ColumnsWithTypeAndName & arguments, const DataTypePtr & res_type, const ColumnNullable * nullable_col, size_t /*input_rows_count*/)
+        {
+            const auto & first_col = arguments.front().column.get();
+            const auto & result_type = typeid_cast<const EnumType &>(*res_type);
+
+            const ColumnNumberType * col = typeid_cast<const ColumnNumberType *>(first_col);
+
+            if (col && nullable_col && nullable_col->size() != col->size())
+                throw Exception(ErrorCodes::LOGICAL_ERROR, "ColumnNullable is not compatible with original");
+
+            if (col)
+            {
+                const auto size = col->size();
+                const auto & in_data = col->getData();
+
+                auto res = result_type.createColumn();
+                auto & out_data = static_cast<typename EnumType::ColumnType &>(*res).getData();
+                out_data.resize(size);
+
+                auto default_enum_value = result_type.getValues().front().second;
+
+                constexpr auto min_value = std::numeric_limits<FieldType>::min();
+                constexpr auto max_value = std::numeric_limits<FieldType>::max();
+                constexpr auto is_signed = is_signed_v<typename ColumnNumberType::ValueType>;
+
+                if (nullable_col)
+                {
+                    for (size_t i = 0; i < size; ++i)
+                    {
+                        if (!nullable_col->isNullAt(i))
```

Comment:
`isNullAt` performs `assert_cast` on each call, better to use `null_map` from `ColumnNullable` directly. Also, we can avoid code duplication on nullable/non-nullable cases. For example, doing smth like:
```
const NullMap * null_map;
if (nullable_col)
    null_map = &nullable_col->getNullMapData();
for (size_t i = 0; i < size; ++i)
{
    if (null_map && (*null_map)[i])
        out_data[i] = default_enum_value;
    else
    {
    ...
    }
}
...
```